### PR TITLE
Bucket aggregation circuit breaker optimization.

### PR DESCRIPTION
**Pull request by howardhuanghua**
Bucket aggregation will consume a lot of memory on coordinate node if it has a huge number of resulting buckets. _search.max_buckets_ setting could limit maximum number of buckets allowed in a single response. Sometimes user may increase this setting to get more buckets, but it also increases the risk of OOM. It's hard to evaluate a suitable value for _max_buckets_.

With this PR, we introduced search.check_buckets_step_size setting. Whenever allocating _search.check_buckets_step_size_ new buckets, we do a parent circuit breaker checking. Based on this setting, user could control aggregation memory with a certain step size. We are also considering whether we could deprecate _search.max_buckets_ in the future.

### CODE REVIEW

**Comment by howardhuanghua** [comment]
Ping @danielmitterdorfer @dakrone , could you please help to check this PR? Thanks.

**Comment by polyfractal** [comment]
Hi @howardhuanghua, thanks for the PR :)

Is this specifically to address https://github.com/elastic/elasticsearch/issues/37182, e.g. dealing with memory usage on the coordinating node (before the final reduce is invoked)?

I'm asking because we do track memory usage with the `request` circuit breaker during the shard collections (as part of `AggregatorBase`), so this breaker seems redundant in that respect.  But if it is to address #37182, then I understand the purpose. 

I'm not sure a new breaker is the right approach though.  It would probably be better to re-use the existing `request` breaker since that is well defined and semantically fits the job.

As an aside, I don't think deprecating `max_buckets` is likely.  MaxBuckets exists because tracking memory overhead of buckets/metrics is very difficult, and so sometimes a hard limit is more robust than attempting to estimate memory usage.

**Comment by howardhuanghua** [comment]
Hi @polyfractal, this PR checks real memory usage with a certain step size of buckets allocation in `MultiBucketConsumer`. The purpose is to break a aggregation request that may get memory explosion as _early_ as possible. **It could be used both in coordinate node and data node**:

- Shard collection phase in `BucketsAggregator` on data node:
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java#L133-L135

- Buckets reduction phase in `InternalAggregation` on coordinate node:
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java#L82-L84

For the aggregation buckets memory tracking issue, now we have follow solutions:

1. Add aggregation buckets memory usage into `request` breaker.
Since aggregation is also a search request, it also should be controlled by `request` circuit breaker as you have suggested in the comment. However, it's difficult to track buckets memory, so it's hard to use `request` circuit breaker to limit buckets memory.

2. Use `max_buckets` to hard limit result buckets number.
This could accurately control buckets number usage, rather than memory usage. It's also not a memory circuit breaker. If user wants to increase it to get more bucket results, it's hard to evaluate a value that would not explode memory and also meet user's result buckets number requirement. Many of our users have some data analysis cases for time series data, the result buckets number may get up to tens or even hundreds of thousands, if we set `max_buckets` directly to a high value for all the aggregation request, it would easily get node OOM. So it's a little bit risky and not that flexible to tune this setting.

3. Check parent circuit breaker in a certain step/interval.
That's the above PR we have introduced, `check_buckets_step_size` controls memory tracking step/interval. Instead of control total result buckets, it uses parent circuit breaker to control memory usage with a certain step size. This would be accurate and flexible, user could set a lower step size to control aggregation memory usage without limiting max result buckets number.
If the purpose of `max_buckets` is only to control memory usage, then I think it seems `check_buckets_step_size` is good enough.

**Comment by polyfractal** [comment]
Thanks for the explanation, that helps me understand the purpose better.  A few thoughts, mainly writing this so I have an overview of all the pieces:

Today, the parent breaker is checked any time we increment a breaker via [`ChildMemoryCircuitBreaker#addEstimateBytesAndMaybeBreak()`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/breaker/ChildMemoryCircuitBreaker.java#L128).  This is used when we create a new Aggregator ([`AggregatorBase#addRequestCircuitBreakerBytes()`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java#L122)) or any time we resize a BigArrays ([`BigArrays#adjustBreaker()`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/util/BigArrays.java#L399)).  BigArrays are the data structure that back most of the aggregations.

So this means the real memory breaker is being checked on most (but not all) shard-level aggregation operation today.

It's _not_ being checked on:

- shard-level reduction
- while accumulating results on the coordinator / intermediate reductions
- the final reduce on coordinator
- response serialization

So I agree that we should add some checks to the parent breaker at those steps.  I'm not sure we need `check_buckets_step_size`  though.  Checking real-memory breaker is pretty cheap ([few hundred nanos](https://github.com/elastic/elasticsearch/pull/31767#issuecomment-402189317)) so I think we could just check every time we account for a new bucket.  That would simplify things since we don't need an extra setting.

@danielmitterdorfer do you know if it is ok to call `HiearchyCircuitBreakingService#checkParentLimit()` directly?  It's public, but also only used from within the service, and seems like maybe something that should only be called indirectly?

**Comment by howardhuanghua** [comment]
Thanks @danielmitterdorfer, @polyfractal. To avoid calling `#checkParentLimit()` directly, I use `REQUEST` child ciruit breaker instead in `MultiBucketConsumer`, pass 0 bytes to `breaker.addEstimateBytesAndMaybeBreak` to trigger parent limit checking.

If we check real-memory breaker every time we account for a new bucket, a single time is about few hundred nanos, if an aggregation result contains hundreds of thousands of buckets, then would cost tens of milliseconds on parent limit checking. With `check_buckets_step_size` setting, we could reduce the checking frequency to avoid performance impaction even it's just a little. I set it to 1000 by default, in above case, even hundreds of thousands of buckets, the checking cost would totally less than 1ms.

Meanwhile, I would like to confirm these unchecked levels you have mentioned above:

- shard-level reduction
Do you mean `AggregatorBase#buildAggregation`? It seems in this shard leve reduction, it already has circuit break checking, take `StringTermsAggregator` as an example:
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsAggregator.java#L154

- while accumulating results on the coordinator / intermediate reductions
We already have a propuse on this. For coordinator node accumulating aggregation results, 

- the final reduce on coordinator
Current PR is going to solve the coordinator node buckets reduction circuit breaker checking issue.

- response serialization
This should be :

**Comment by howardhuanghua** [comment]
Thanks @danielmitterdorfer, @polyfractal. To avoid calling `#checkParentLimit()` directly, I use `REQUEST` child ciruit breaker instead in `MultiBucketConsumer`, pass 0 bytes to `breaker.addEstimateBytesAndMaybeBreak` to trigger parent limit checking. Please help to check the updated commit.

If we check real-memory breaker every time we account for a new bucket, each checking is about few hundred nanos, if an aggregation result contains hundreds of thousands of buckets, then would cost tens of milliseconds on parent limit checking. With `check_buckets_step_size` setting, we could reduce the checking frequency to avoid performance impaction even it's just a little. I set it to `1000` by default, in above case, even hundreds of thousands of buckets, the checking cost would totally less than 1ms.

Meanwhile, I would like to confirm these unchecked levels you have mentioned above:

- shard-level reduction
Do you mean `AggregatorBase#buildAggregation`? It seems in this shard leve reduction, it already has circuit break checking, take `StringTermsAggregator` as an example:
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsAggregator.java#L154

- while accumulating results on the coordinator / intermediate reductions
We already have a propose on this. For coordinator node accumulating aggregation results, we could evaluate buckets memory usage based on incoming network stream bytes length before deserialization, and add it into request circuit breaker. I will open another PR to describe this propose later.

- the final reduce on coordinator
Current PR could solve the coordinator node buckets reduction circuit breaker checking issue.

- response serialization
Do you mean the search response building method as follow?
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java#L299-L301

**Comment by howardhuanghua** [comment]
Hi @polyfractal, would you pelase help to check the updated commit again? Thanks a lot.

**Comment by polyfractal** [comment]
Hi @howardhuanghua, apologies for the delay.  I became unexpectedly very busy this last week.  I'll try to take a look at your new changes tomorrow!

**Comment by howardhuanghua** [comment]
Hi @polyfractal, I opened another PR https://github.com/elastic/elasticsearch/pull/47806 for memory checking on accumulating results on the coordinator. Please help to check if have you free time. Thank you.

**Comment by howardhuanghua** [comment]
Hi Adrien @jpountz , would you please help to review this PR? Thanks a lot.

File: server/src/main/java/org/elasticsearch/search/aggregations/MultiBucketConsumerService.java [hunk 1]
```diff
@@ -109,6 +129,11 @@ public void accept(int value) {
                     + "] but was [" + count + "]. This limit can be set by changing the [" +
                     MAX_BUCKET_SETTING.getKey() + "] cluster level setting.", limit);
             }
+
+            if (value > 0 && checkBucketsStepSizeLimit > 0 && count % checkBucketsStepSizeLimit == 0) {
+                CircuitBreaker breaker = circuitBreakerService.getBreaker(CircuitBreaker.REQUEST);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/MultiBucketConsumerService.java]: Instead of only getting the request circuit breaker here, can you take a CircuitBreaker instead of a CircuitBreakerService in the constructor?

File: server/src/main/java/org/elasticsearch/search/aggregations/MultiBucketConsumerService.java [hunk 2]
```diff
@@ -38,14 +40,27 @@
...
 public class MultiBucketConsumerService {
     public static final int DEFAULT_MAX_BUCKETS = 10000;
+    public static final int DEFAULT_CHECK_BUCKETS_STEP_SIZE = 1000;
     public static final Setting<Integer> MAX_BUCKET_SETTING =
         Setting.intSetting("search.max_buckets", DEFAULT_MAX_BUCKETS, 0, Setting.Property.NodeScope, Setting.Property.Dynamic);
 
-    private volatile int maxBucket;
+    public static final Setting<Integer> CHECK_BUCKETS_STEP_SIZE_SETTING =
+        Setting.intSetting("search.check_buckets_step_size", DEFAULT_CHECK_BUCKETS_STEP_SIZE,
+            -1, Setting.Property.NodeScope, Setting.Property.Dynamic);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/MultiBucketConsumerService.java]: I'd be in favor of not making it configurable at all and check every 1000 buckets all the time? (Or maybe 1024 so that the `% 1000` can be replaced with a lighter `& 0x3FF` mask)

File: test/framework/src/main/java/org/elasticsearch/node/MockNode.java [hunk 3]
```diff
@@ -72,27 +72,27 @@ public MockNode(final Settings settings, final Collection<Class<? extends Plugin
...
+            forbidPrivateIndexSettings);
     }
 
     private MockNode(
-            final Environment environment,
-            final Collection<Class<? extends Plugin>> classpathPlugins,
-            final boolean forbidPrivateIndexSettings) {
+        final Environment environment,
+        final Collection<Class<? extends Plugin>> classpathPlugins,
+        final boolean forbidPrivateIndexSettings) {
```
Comment (jpountz) [inline test/framework/src/main/java/org/elasticsearch/node/MockNode.java]: can you undo the above unrelated indentation changes?

File: server/src/test/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerServiceTests.java [hunk 4]
```diff
@@ -293,6 +294,31 @@ public void testTrippedCircuitBreakerDurability() {
...
+            new ClusterSettings(clusterSettings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS))) {
+
+            long parentLimitBytes = ((HierarchyCircuitBreakerService) service).getParentLimit();
+            assertEquals(new ByteSizeValue(20, ByteSizeUnit.MB).getBytes(), parentLimitBytes);
+
+            MultiBucketConsumerService.MultiBucketConsumer multiBucketConsumer =
+                new MultiBucketConsumerService.MultiBucketConsumer(10000, 10000, service);
+
+            long currentMemory = ((HierarchyCircuitBreakerService) service).currentMemoryUsage();
+            if (currentMemory > parentLimitBytes) {
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerServiceTests.java]: Can you maybe make the test a bit more predictable by calling `addWithoutBreaking` with a number of bytes that is greater than the limit?

File: test/framework/src/main/java/org/elasticsearch/search/MockSearchService.java [hunk 5]
```diff
@@ -67,9 +68,9 @@ static void removeActiveContext(SearchContext context) {
     }
 
     public MockSearchService(ClusterService clusterService,
-            IndicesService indicesService, ThreadPool threadPool, ScriptService scriptService,
-            BigArrays bigArrays, FetchPhase fetchPhase) {
-        super(clusterService, indicesService, threadPool, scriptService, bigArrays, fetchPhase, null);
+                             IndicesService indicesService, ThreadPool threadPool, ScriptService scriptService,
+                             BigArrays bigArrays, FetchPhase fetchPhase, CircuitBreakerService circuitBreakerService) {
```
Comment (jpountz) [inline test/framework/src/main/java/org/elasticsearch/search/MockSearchService.java]: same here

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 6]
```diff
@@ -254,12 +255,12 @@ public boolean shouldCache(Query query) {
...
-                        .numberOfReplicas(0)
-                        .creationDate(System.currentTimeMillis())
-                        .build(),
-                Settings.EMPTY
+            IndexMetaData.builder("_index").settings(Settings.builder().put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT))
+                .numberOfShards(1)
+                .numberOfReplicas(0)
+                .creationDate(System.currentTimeMillis())
+                .build(),
+            Settings.EMPTY
```
Comment (jpountz) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java]: same here

**Review by jpountz — COMMENTED** [review]
I like the idea a lot. I left some comments that might help simplify a bit.

**Comment by howardhuanghua** [comment]
Hi @jpountz , thanks for your review. I have updated the code, would you please help to check again?
Add few points as follow:
1. I removed `search.check_buckets_step_size` that I've added before. I agree that we don't need user to configure this. Meanwhile, I am considering that, **shall we still need `search.max_buckets` ?** It seems if we check parent circuit breaker every 1024 buckets, then we don't need to limit max buckets anymore?

2. In `testAllocationBucketsBreaker` UT, I use `addWithoutBreaking` to make sure used bytes is greater than the total circuit breaker limit, and I also need to set `indices.breaker.total.use_real_memory` to `false`. As using real memory case would not add all the child breakers together.

**Comment by howardhuanghua** [comment]
Hi @jpountz , would you please help to review the updated commit again? Thank you.

**Comment by howardhuanghua** [comment]
Hi @jpountz  , would please help to review the changes again? Thank you. :)

**Comment by howardhuanghua** [comment]
Hi @jpountz, would please help to review the changes again? Thank you.

**Comment by howardhuanghua** [comment]
Hi @jpountz , would you please help to review the changes again? Thank you.

**Comment by jpountz** [comment]
@elasticmachine update master

**Comment by jpountz** [comment]
@elasticmachine update branch

**Comment by jpountz** [comment]
@elasticmachine ok to test

**Comment by jpountz** [comment]
I opened https://github.com/elastic/elasticsearch/issues/51694 for the failure of elasticsearch-ci/1

@elasticmachine run elasticsearch-ci/1

**Comment by jpountz** [comment]
@howardhuanghua Thanks again for your contribution! I opened https://github.com/elastic/elasticsearch/issues/51731 to discuss the deprecation of `search.max_buckets` that you raised.

**Comment by howardhuanghua** [comment]
@jpountz Thanks a lot for your help!

**Comment by polyfractal** [comment]
Thanks @howardhuanghua! :heart:

**Comment by liwanjie1020** [comment]
@howardhuanghua I always thought that with this new parameter, requests for too many buckets should be rejected。But I tested it today, making two requests and creating hundreds of thousands of buckets per request。In the end, the machine is still directly OOM。
My ES version is 7.10.1。I tested not only the case where indices.breaker.total.use_real_memory is false, but also the case where indices.breaker.total.use_real_memory is true。And the max_buckets is the default value of 65535.The result is the same。
image
image
But when I set the max_buckets to extremely large (instead of the default value of 65535). The circuit breaker will come into play and the cluster will not crash and will not OOM.
I feel very confused. I hope you can help me. Thank you very much

**Comment by iWingsChen** [comment]
If the JVM memory has been used for 60%, a large request comes in and consumes 50% of the memory. Since the default value of the circuit breaker is 60%, the circuit breaker will not be triggered, and the consequence is oom
